integrations/sonarqube/client.py

```diff
@@ -468,6 +472,49 @@ async def get_all_sonarqube_analyses(
             )
             yield analysis_data
 
+    async def _get_all_portfolios(self) -> list[dict[str, Any]]:
+        logger.info(
+            f"Fetching all root portfolios in organization: {self.organization_id}"
+        )
+        response = await self.send_api_request(endpoint=Endpoints.PORTFOLIOS)
+        return response.get("views", [])
+
+    async def get_portfolio_details(self, portfolio_key: str) -> dict[str, Any]:
+        logger.info(f"Fetching portfolio details for: {portfolio_key}")
+        response = await self.send_api_request(
+            endpoint=Endpoints.PORTFOLIO_DETAILS,
+            query_params={"key": portfolio_key},
+        )
+        return response
+
+    def _extract_subportfolios(self, portfolio: dict[str, Any]) -> list[dict[str, Any]]:
+        logger.info(f"Fetching subportfolios for: {portfolio['key']}")
+        subportfolios = portfolio.get("subViews", []) or []
+        all_portfolios = []
+        for subportfolio in subportfolios:
+            if subportfolio.get("qualifier") in PORTFOLIO_VIEW_QUALIFIERS:
+                all_portfolios.append(subportfolio)
+            all_portfolios.extend(self._extract_subportfolios(subportfolio))
+        return all_portfolios
+
+    async def get_all_portfolios(self) -> AsyncGenerator[list[dict[str, Any]], None]:
+        logger.info(f"Fetching all portfolios in organization: {self.organization_id}")
+        portfolios = await self._get_all_portfolios()
+
+        try:
+            portfolio_keys = [portfolio["key"] for portfolio in portfolios]
+            portfolios_data = await asyncio.gather(
```

Comment:
Let's batch concurrent requests to prevent the integration from running possibly thousands of tasks concurrently